### PR TITLE
Migrated pkg/controller/job to contextual logging

### PR DESCRIPTION
**Pull request by fatsheep9146**
Signed-off-by: Ziqi Zhao <zhaoziqi9146@gmail.com>

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Use contextual logging for pkg/controller/job

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/enhancements/issues/3077

#### Does this PR introduce a user-facing change?

```release-note
Migrated the Job controller (within `kube-controller-manager`) to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging
```

### CODE REVIEW

**Comment by fatsheep9146** [comment]
@pohly could you help review this

**Comment by kannon92** [comment]
Hello @fatsheep9146, it looks like your PR has a real failure in the linter.

```
pkg/controller/job/job_controller.go:1521:5: logging function "Info" should not use format specifier "%d" (logcheck)
				logger.V(2).Info("Slow-start failure. Skipping creation of %d pods, decrementing expectations", skippedPods, "job", klog.KObj(job))
				^
```

File: pkg/controller/job/job_controller.go [hunk 1]
```diff
@@ -1493,7 +1507,7 @@ func (jm *Controller) manageJob(ctx context.Context, job *batch.Job, activePods
 					if err != nil {
 						defer utilruntime.HandleError(err)
 						// Decrement the expected number of creates because the informer won't observe this pod
-						klog.V(2).Infof("Failed creation, decrementing expectations for job %q/%q", job.Namespace, job.Name)
+						logger.V(2).Info("Failed creation, decrementing expectations", "job", klog.KObj(job))
```
Comment (pohly) [inline pkg/controller/job/job_controller.go:1510]: The key "job" is used inconsistently: above it refers to a string, here to an object.

I think I prefer to use "job" for the object and "key" when referring to the string which identifies the work item from the queue.

File: pkg/controller/job/job_controller_test.go [hunk 2]
```diff
@@ -3380,8 +3396,10 @@ func TestAddPod(t *testing.T) {
 }
 
 func TestAddPodOrphan(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
+	logger := klog.FromContext(ctx)
```
Comment (pohly) [inline pkg/controller/job/job_controller_test.go:3400]:
```suggestion
	logger, ctx := ktesting.NewTestContext(t)
```

File: pkg/controller/job/job_controller_test.go [hunk 3]
```diff
@@ -3461,8 +3481,10 @@ func TestUpdatePod(t *testing.T) {
 }
 
 func TestUpdatePodOrphanWithNewLabels(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
+	logger := klog.FromContext(ctx)
```
Comment (pohly) [inline pkg/controller/job/job_controller_test.go:3485]:
```suggestion
	logger, ctx := ktesting.NewTestContext(t)
```

File: pkg/controller/job/job_controller_test.go [hunk 4]
```diff
@@ -3539,15 +3565,17 @@ func TestUpdatePodRelease(t *testing.T) {
...
-	jm.updatePod(&prev, pod1)
+	jm.updatePod(logger, &prev, pod1)
 	if got, want := jm.queue.Len(), 2; got != want {
 		t.Fatalf("queue.Len() = %v, want %v", got, want)
 	}
 }
 
 func TestDeletePod(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
+	logger := klog.FromContext(ctx)
```
Comment (pohly) [inline pkg/controller/job/job_controller_test.go:3576]:
```suggestion
	logger, ctx := ktesting.NewTestContext(t)
```

**Review by pohly — CHANGES_REQUESTED** [review]

**Review by pohly — CHANGES_REQUESTED** [review]
Looks good now. Let's clean up and then it is ready:
- remove the exception from https://github.com/kubernetes/kubernetes/blob/5f23f835879b514e9bf16809059b38957d20a486/hack/logcheck.conf#L49
- squash all commits into one, without rebasing

**Comment by fatsheep9146** [comment]
> Looks good now. Let's clean up and then it is ready:
> 
> * remove the exception from https://github.com/kubernetes/kubernetes/blob/5f23f835879b514e9bf16809059b38957d20a486/hack/logcheck.conf#L49
> * squash all commits into one, without rebasing

Done

**Comment by pohly** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/116910/pull-kubernetes-verify/1646077352406421504 shows that there are klog calls left that still need to be converted in this PR.

**Comment by fatsheep9146** [comment]
@pohly could you check this again?

**Review by pohly — APPROVED** [review]
/lgtm
/assign @soltysh 

For approval.

**Comment by fatsheep9146** [comment]
@pohly could you help add lgtm label again?

**Comment by fatsheep9146** [comment]
@pohly Could we push this pr into merged ASAP? The pr is already added with lgtm label before.

**Review by pohly — APPROVED** [review]
/lgtm

**Review by soltysh — APPROVED** [review]
/approve

**Pull request merged**